ARROW-15037: [C#]  A Real-time memory-based data analytics example

zixi-bwang opened the pull request:
A Real-time C# memory-based data analytics platform, which leverages Apache Arrow as the unified data store.

zixi-bwang commented:
Hi, @eerhardt 

I added a usage example, could you review?

eerhardt commented:
One quick question/suggestion: Can you add a small `README` to the example explaining what it is showing? Describe the scenario, etc.

kou commented:
I think that this is not a MINOR change. See our MINOR definition: https://github.com/apache/arrow/blob/master/CONTRIBUTING.md#Minor-Fixes

> Any functionality change should have a JIRA opened. For minor changes that affect documentation, you do not need to open up a JIRA. Instead you can prefix the title of your PR with "MINOR: " if meets the following guidelines:
>
>  * Grammar, usage and spelling fixes that affect no more than 2 files
>  * Documentation updates affecting no more than 2 files and not more than 500 words.

Could you open an issue on JIRA? https://issues.apache.org/jira/browse/ARROW

See also the auto-generated comment: https://github.com/apache/arrow/pull/11874#issuecomment-987128831

zixi-bwang commented:
> One quick question/suggestion: Can you add a small `README` to the example explaining what it is showing? Describe the scenario, etc.

Hi @eerhardt, I added the readme also updated the example, please review.

eerhardt commented on csharp/examples/IoTExample/IoTExample.csproj:
```diff
@@ -0,0 +1,16 @@
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
```
```suggestion
    <TargetFramework>netcoreapp3.1</TargetFramework>
```

2.1 is EOL'd. Can you update this to 3.1?

eerhardt commented on csharp/examples/IoTExample/README.md:
```diff
@@ -0,0 +1,29 @@
+Each participant performed each of the 18 activities for a total amount of one billion accelerometer data events
+reported from smartphone and smartwatch.
+
+Timestamp is the time at which the sensor reported the reading.
+X_Axis is the g-force acceleration along the x-axis.
+Y_Axis is the g-force acceleration along the y-axis.
+Z_Axis is the g-force acceleration along the z-axis.
```
```suggestion
* Timestamp is the time at which the sensor reported the reading.
* X_Axis is the g-force acceleration along the x-axis.
* Y_Axis is the g-force acceleration along the y-axis.
* Z_Axis is the g-force acceleration along the z-axis.
```

It may make this easier to read with bulleting.

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataset.cs:
```diff
@@ -0,0 +1,211 @@
+
+            Schema schema = _recordBatches[0].Schema;
+            // Write record batch to a file
+            using (var stream = File.OpenWrite(@"c:\temp\data\iotbigdata.arrow"))
```
If this path doesn't exist, you get an error at the end of the run:

```
Producing is blocked, percent completed is: 90%
Unhandled exception. System.IO.DirectoryNotFoundException: Could not find a part of the path 'c:\temp\data\iotbigdata.arrow'.
   at System.IO.FileStream.ValidateFileHandle(SafeFileHandle fileHandle)
   at System.IO.FileStream.CreateFileOpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at System.IO.File.OpenWrite(String path)
   at IoTExample.SampleDataset.PersistData() in C:\git\arrow\csharp\examples\IoTExample\Model\SampleDataset.cs:line 185
   at IoTExample.Program.Main(String[] args) in C:\git\arrow\csharp\examples\IoTExample\Program.cs:line 42
   at IoTExample.Program.<Main>(String[] args)
```

Maybe just write the file to the current directory?

zixi-bwang replied:
Updated.

eerhardt commented on csharp/examples/IoTExample/Program.cs:
```diff
@@ -0,0 +1,71 @@
+                return;
+
+            Console.WriteLine("Reading arrow files...");
+            var stream = File.OpenRead(@"c:\temp\data\iotbigdata.arrow");
```
Since this path needs to be in sync with `PersistData`, it may make sense to declare it in one spot, and pass it into `SampleDataset`.

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataset.cs:
```diff
@@ -0,0 +1,211 @@
+            _rows.CompleteAdding();
+        }
+
+        public async void Consume()
```
This method doesn't need to be `async` since it doesn't `await` anything.

Also - avoid `async void` methods in general. Always return `async Task`, if you can.

zixi-bwang replied:
Done.

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataset.cs:
```diff
@@ -0,0 +1,211 @@
+        private int _size;
+        private readonly int _inputs;
+        private readonly int _capacity;
+        private readonly BlockingCollection<SensorData> _rows;
```
Maybe since we are using `async` methods, can we change this to use [Channels](https://devblogs.microsoft.com/dotnet/an-introduction-to-system-threading-channels/) instead? They are very similar, but Channels was made with `async` in mind.

zixi-bwang replied:
Sure, I'll try to use Channel.

zixi-bwang replied:
The code has been refactored, please review.

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataset.cs:
```diff
@@ -0,0 +1,211 @@
+
+        public async Task<bool> PersistData()
+        {
+            string time = DateTime.Now.ToString("yyyyMMdd_HHmmss");
```
```suggestion
```

This variable doesn't appear to be used.

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataset.cs:
```diff
@@ -0,0 +1,211 @@
+            }
+        }
+
+        public async Task<bool> PersistData()
```
```suggestion
        public async Task PersistData()
```

This method never returns `false`. So you can just change it to return `Task` instead.

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataset.cs:
```diff
@@ -0,0 +1,211 @@
+                }
+                else
+                {
+                    //Console.WriteLine($"Dequeue Task 1");
```
```suggestion
```

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataset.cs:
```diff
@@ -0,0 +1,211 @@
+
+                if (success)
+                {
+                    //Console.WriteLine($"Enqueue Task 0");
```
```suggestion
```

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataPipeline.cs:
```diff
@@ -0,0 +1,197 @@
+        public SampleDataPipeline(int concurrencyLevel, int totalSensorData, int queueCapacity)
+        {
+            _partitions = concurrencyLevel;
+            _inputs = totalSensorData;
```
I find it confusing to use two different names here. Can we pick one and use it both for the field and the parameter?

zixi-bwang replied:
Done.

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataPipeline.cs:
```diff
@@ -0,0 +1,197 @@
+
+            for (int i = 0; i < _partitions; i++)
+            {
+                _colSubjectIdArrays.Add(new ConcurrentBag<int>());
```
ConcurrentBag is unordered. Meaning when you enumerate it, you can get the results in a different order than what they went in. Is that intended?

https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentbag-1?view=net-6.0

zixi-bwang replied:
My intention was to find a data structure that has below characteristics:

1. Inserting by row
2. Querying by column
3. The "results in a different order than what they went in" does not matter, but the kth item in one column should belong to the same row as the kth item in another column
4. So all of the column arrays can be appended to a RecordBatch

It seems DataTable is a good candidate, or could you recommend other light weight data structures instead?

eerhardt replied:
> but the kth item in one column should belong to the same row as the kth item in another column

Since a "bag" is unordered, there is no such thing as the "kth item". Notice there is no indexer. You can enumerate the items, but you aren't guaranteed the order they are returned.

> It seems DataTable is a good candidate, or could you recommend other light weight data structures instead?

I wouldn't use DataTable. Instead, I would eliminate the concurrency in `ReadFromChannel`. Have many "writer" threads all writing items to the channel. But then only have 1 "reader" thread that pulls items off the channel and writes them directly to a `RecordBatch.Builder`.

zixi-bwang replied:
I have refactored the code per your suggestion to use multiple writers and 1 reader only, and writes data directly to builder.
And I also modified the example to emulate below scenario:

- There are around 9,000 sensors(subjects) in total
- Each sensor can send hundreds or thousands events during a short time of period
- When reading from channel, data is appended to each array builder hashed by SubjectId as the key of builderId
- When building RecordBatch, the array builders with the same builderId are built together
- The final Arrow file contains 9,000 RecordBatches 

Do you think this a good scenario for real world applications in terms of utilizing Arrow RecordBatch?

eerhardt commented on csharp/examples/IoTExample/Model/SampleDataPipeline.cs:
```diff
@@ -0,0 +1,197 @@
+        private int _size;
+        private int _partitions;
+        private readonly int _inputs;
+        private readonly int _capacity;
```
Is this field necessary?

zixi-bwang replied:
Not really, removed.

eerhardt commented:
@zixi-bwang - some bad merge must have happened. This PR is now affecting A LOT of files. Can you fix it?

zixi-bwang commented:
> @zixi-bwang - some bad merge must have happened. This PR is now affecting A LOT of files. Can you fix it?

Not sure how to fix it, I'll have to close the PR then.
